src/ucp/core/ucp_ep.c

```diff
@@ -2744,7 +2744,11 @@ static void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
     ucp_trace_req(req, "purged with status %s (%d) on ep %p",
                   ucs_status_string(status), status, ucp_ep);
 
-    ucp_request_id_release(req);
+    /* Only send operations could have request ID allocated */
+    if (!(req->flags &
+          (UCP_REQUEST_FLAG_RECV_AM | UCP_REQUEST_FLAG_RECV_TAG))) {
+        ucp_request_id_release(req);
```

Comment:
if ucp_request_id_release can be used only for send requests, we should rename it (and also all funcs which use req->send) to ucp_send_request_id_release
maybe even add request debug flag "SEND" and check it

Reply:
maybe it is worth to add EP as a parameter to those functions and be able to use it for all operations (SEND an RECEIVE), since ID is a common part of request

Reply:
nevermind.
your idea is better - done

Reply:
but adding `SEND` flag isn't needed, we can check for `RECV_AM`/`RECV_TAG`